crates/chia-consensus/src/merkle_tree.rs

```diff
@@ -0,0 +1,1362 @@
+// This file contains the code used to create a full MerkleSet and is heavily reliant on the code in merkle_set.rs.
+
+#[cfg(test)]
+use crate::merkle_set::compute_merkle_set_root;
+use crate::merkle_set::{get_bit, hash, NodeType, BLANK};
+use clvmr::sha2::{Digest, Sha256};
+#[cfg(feature = "py-bindings")]
+use pyo3::exceptions;
+#[cfg(feature = "py-bindings")]
+use pyo3::prelude::*;
+#[cfg(feature = "py-bindings")]
+use pyo3::types::{PyBytes, PyList};
+#[cfg(feature = "py-bindings")]
+use pyo3::{pyclass, pymethods, PyResult};
+#[cfg(test)]
+use rand::rngs::SmallRng; // cargo says this isn't required but tests won't run without it
+#[cfg(test)]
+use rand::{Rng, SeedableRng}; // needed for PyBytes
+
+// the ArrayType is used to create a more lasting MerkleTree representation in the MerkleSet struct
+#[derive(PartialEq, Debug, Copy, Clone)]
+pub enum ArrayTypes {
+    Leaf { data: usize },                // indexes for the data_hash array
+    Middle { children: (usize, usize) }, // indexes for a Vec of ArrayTypes
+    Empty,
+    Truncated,
+}
+
+// represents a MerkleTree by putting all the nodes in a vec. Root is the last entry.
+#[derive(PartialEq, Debug, Clone)]
+#[cfg_attr(feature = "py-bindings", pyclass(frozen, name = "MerkleSet"))]
+pub struct MerkleSet {
+    pub nodes_vec: Vec<ArrayTypes>,
+    pub leaf_vec: Vec<[u8; 32]>,
+    pub hash_cache: Vec<[u8; 32]>, // same size and order as nodes_vec
+}
+
+const EMPTY: u8 = 0;
+const TERMINAL: u8 = 1;
+const TRUNCATED: u8 = 2;
+const MIDDLE: u8 = 3;
+
+const EMPTY_NODE_HASH: [u8; 32] = [
+    127, 156, 158, 49, 172, 130, 86, 202, 47, 37, 133, 131, 223, 38, 45, 188, 125, 111, 104, 242,
+    160, 48, 67, 213, 201, 154, 74, 229, 167, 57, 108, 233,
+];
+
+#[derive(Debug)]
+#[cfg_attr(feature = "py-bindings", pyclass(frozen, name = "SetError"))]
+pub struct SetError;
+
+pub fn deserialize_proof(proof: &[u8]) -> Result<MerkleSet, SetError> {
```

Comment:
this function should have a fuzzer

Reply:
a fuzzer is much better at finding edge cases. It's not random, it's code coverage guided.

Reply:
I commented in the tests; they pass in valid proofs (or valid but truncated proofs). when we deserialize proofs from the network, we can't trust them so we need a fuzzer to test potentially malicious input.